gdc_rnaseq_tools/augment_star_counts.py

```diff
@@ -0,0 +1,228 @@
+import numpy as np
+import pandas as pd
+
+from gdc_rnaseq_tools.utils import DataError, DataFormatError, get_logger
+
+
+def load_counts_table(counts):
+    '''
+    Loads per-gene counts data table as created by STAR
+
+
+    returns pandas DataFrame with the following columns:
+        gene_id
+        unstranded
+        stranded_first
+        stranded_second
+    '''
+    cdf = pd.read_table(
+        counts,
+        header=None,
+        names=['gene_id', 'unstranded', 'stranded_first', 'stranded_second'],
+    )
+    return cdf
+
+
+def load_gene_info(gene_info):
+    '''
+    Loads table with additional information used to either annotate or 
+    calculate values needed to augment counts table.
+
+    file should be TSV and should have the following columns:
+        gene_id
+        total_exon_length
+        gene_name
+        gene_type
+        Chromosome
+
+    returns pandas DataFrame with all columns
+    '''
+    # load data
+    gdf = pd.read_table(gene_info)
+
+    # check that data has exactly the columns expected
+    expected_cols = [
+        'gene_id',
+        'total_exon_length',
+        'gene_name',
+        'gene_type',
+        'Chromosome',
+    ]
+    if set(expected_cols) - set(gdf.columns) != set():
+        raise DataFormatError('Expected columns not found')
+
+    # would be good to check that file is the correct length, but that's
+    # tied to the annotation and I'm not sure how to do that
+    return gdf
+
+
+def prep_data(counts, gene_info):
+    '''
+    Load data files and merge on gene_id via "inner" join operation.
+
+    STAR counts have 4 extra lines at the top reporting unmapped,
+    multimapping, noFeature, and ambiguous reads. These are not included
+    in the merged DataFrame, but are returned in the extras DataFrame to
+    be merged back into the final output.
+
+    Returns a tuple of:
+        merged DataFrame
+        extras DataFrame
+    '''
+    cdf = load_counts_table(counts)
+    gdf = load_gene_info(gene_info)
+    df = pd.merge(gdf, cdf, on='gene_id', how='inner')
+
+    if len(df) != len(gdf):
+        raise DataError(
+            'Data length or key mismatch between gene counts and gene '
+            + 'annotation tables. Possible file truncation.'
+        )
+
+    # save table of mis-aligned read stats also reported by star
+    mdf = cdf.iloc[0:4].copy()
+
+    return df, mdf
+
+
+def calc_tpm(df, excol='unstranded', lencol='total_exon_length'):
```

Comment:
Here it might be easier to update this method to accept the columns:
```python
def calculate_tpm(unstranded_column, total_length_column):
    rpk = unstranded_column * 1e3 / total_length_column
```
This way, the code is independent of the actual column names, in case those change.